### PR TITLE
Angular/React: Migrates FolderPicker from angular to react

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
Migrates the FolderPicker component from angular to react as part of the ongoing migration process.

**Special notes for your reviewer**:
This is just a simple migration, so no major changes in behavior, though I think at some point it might be worthwhile to investigate if some UX improvements could be made. @jessover9000 

Also vscode seems to have messed with the formatting of the html files - hopefully not too big an issue though? 😅

### CODE REVIEW

File: public/app/features/dashboard/components/FolderPicker/FolderPicker.tsx [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        const isPersistedDashBoard = dashboardId ? true : false;
+        if (isPersistedDashBoard) {
+          folder = resetFolder;
+        } else {
+          folder = options.length > 0 ? options[0] : resetFolder;
+        }
+      }
+    }
+
+    this.setState({
```
Comment (dprokop) [inline public/app/features/dashboard/components/FolderPicker/FolderPicker.tsx]: :( async set state problem strikes again :(

**Comment by marefr** [comment]
Would this allow to retrieve `folderUid` and/or the full folder payload in the select event? Had an [issue](21019) opened 2 days ago where the implementation of that would need to store `folderUid` instead of `folderId` in dashboard list panel.

File: public/app/features/dashboard/components/FolderPicker/FolderPicker.tsx [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    newFolderName: '',
+    createNewFolder: false,
+    validationError: '',
+    hasValidationError: false,
+    newFolderNameTouched: false,
+    isLoading: true,
+  };
+
+  componentDidMount = async () => {
+    await this.loadInitialValue();
```
Comment (hugohaggmark) [inline public/app/features/dashboard/components/FolderPicker/FolderPicker.tsx]: Maybe consider moving all async operations to Redux action and move state to Redux state?

**Review by hugohaggmark — COMMENTED** [review]
Great work on pushin these migrations! I left some a comment that you might want to consider

**Comment by dprokop** [comment]
@kaydelaney do you plan to work on this in the near future?

File: packages/grafana-ui/src/components/Forms/Select/getSelectStyles.ts [hunk 3]
```diff
@@ -16,6 +16,7 @@ export const getSelectStyles = stylesFactory((theme: GrafanaTheme) => {
       box-shadow: 0px 4px 4px ${menuShadowColor};
       position: absolute;
       min-width: 100%;
+      z-index: 1;
```
Comment (peterholmberg) [inline packages/grafana-ui/src/components/Forms/Select/getSelectStyles.ts]: Adding this to fix issue with overlapping menu.

**Comment by dprokop** [comment]
Good stuff.

Couple of issues found:
1. Dashboard settings: the default value of the folder is not there. Select has an empty value
2. Save dashboard modal: default folder is not picked. Seems like the problem applies to all folder pickers
3. Folder picker select does not allow blur (reported in a separate https://github.com/grafana/grafana/issues/21867)
4. Querying for search is not debounced - every keystroke = query
5. This introduces inconsistency in the forms, as it's using new styles. We need to migrate the affected forms to new styles in first order after this is merged

File: public/app/core/angular_wrappers.ts [hunk 4]
```diff
@@ -23,6 +23,7 @@ import ReactProfileWrapper from 'app/features/profile/ReactProfileWrapper';
 import { LokiAnnotationsQueryEditor } from '../plugins/datasource/loki/components/AnnotationsQueryEditor';
 import { HelpModal } from './components/help/HelpModal';
 import { Footer } from './components/Footer/Footer';
+import { FolderPicker } from 'app/core/components/Select/FolderPicker';
```
Comment (dprokop) [inline public/app/core/angular_wrappers.ts]: Hmm, I would suggest `app/core/components/FolderPicker`

File: public/app/core/components/Select/FolderPicker.tsx [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+    return options;
+  };
+
+  onFolderChange = async (newFolder: SelectableValue<number>) => {
+    if (!newFolder) {
+      newFolder = { value: 0, label: this.props.rootName };
+    }
+
+    if (newFolder.__isNew__) {
```
Comment (dprokop) [inline public/app/core/components/Select/FolderPicker.tsx]: Please update to approach from #21869

File: public/app/core/angular_wrappers.ts (hunk 4 above)
Comment (peterholmberg) [reply public/app/core/angular_wrappers.ts]: we have a bunch of different "pickers" under /Select though

**Comment by dprokop** [comment]
Something wrong happened with merge/rebase @peterholmberg ?

File: public/app/core/components/Select/FolderPicker.test.tsx [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+jest.mock('app/core/core', () => ({
+  contextSrv: {
+    isEditor: true,
+  },
+}));
+
+describe('FolderPicker', () => {
+  it('should render', () => {
+    const wrapper = shallow(<FolderPicker onChange={jest.fn()} />);
+    expect(wrapper).toMatchSnapshot();
```
Comment (dprokop) [inline public/app/core/components/Select/FolderPicker.test.tsx]: Let's remove snapshot and just check if the render doesn't fail

File: public/app/core/angular_wrappers.ts (hunk 4 above)
Comment (dprokop) [reply public/app/core/angular_wrappers.ts]: Ah, this is core, you are right

File: public/app/core/components/Select/FolderPicker.tsx [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    );
+  };
+
+  render() {
+    const { folder, validationError, hasValidationError } = this.state;
+    const { enableCreateNew } = this.props;
+
+    return (
+      <>
+        <div className="gf-form-inline">
```
Comment (dprokop) [inline public/app/core/components/Select/FolderPicker.tsx]: Next step here is to replace the legacy form styles with `<Forms.Field` component. Let's do it when migrating views with Folder Picker applied.

**Review by dprokop — APPROVED** [review]
Works good. The twisted select value selection logic requires a separate PR to be refactored though.

**Pull request merged**